### PR TITLE
Implement Users API for user authentication.

### PR DESCRIPTION
**Pull request by zrideaux**
+ Update User.java to include `createUserEntity` method.
+ Create Authentication.java servlet.
  + Generate authentication details when requested:
    + Sign in/out links
    + Logged in boolean
    + Admin status
    + Entity status (more below)
  + Handle new user entity creation and datastore placement for first time logins.
+ Prepare CreateListing.java for authentication check and future use of `addListingKeyToUserEntity` method (likely the PR to follow the updates to the HTML files explained below).
+ Update AuthenticationUtility.java methods to be static and add new `getCurrentUserEntity` method.
+ Move old authentication.js code to authentication-legacy.js.
+ Rework authentication.js to work with Users API.
  + NOTE: I do not plan on merging this change right now. I will swap the authentication scripts in the HTML files after talking with the team before doing so. I also expect this to be fairly straightforward, so don't worry too much :-)
  + Includes general use `authenticate` method with example comments.
***
While logged out:
Screenshot 2020-08-04 at 5 59 04 PM

Local login screen:
Screenshot 2020-08-04 at 5 59 58 PM

While logged in:
Screenshot 2020-08-04 at 6 00 11 PM

### CODE REVIEW

File: src/main/java/com/google/sps/utility/AuthenticationUtility.java [hunk 1]
```diff
@@ -113,11 +113,12 @@ public boolean userAlreadyHasAccount(DatastoreService datastore, String email) {
   /**
    * Retrieves the User entity associated with the currently logged in user
    * if one exists.
+   * @depracated
```
Comment (J-coder13) [inline src/main/java/com/google/sps/utility/AuthenticationUtility.java:116]: What does this mean?

File: src/main/java/com/google/sps/servlets/CreateListing.java [hunk 2]
```diff
@@ -17,14 +17,18 @@
...
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
+import com.google.sps.data.User;
+import com.google.sps.utility.AuthenticationUtility;
 import com.google.sps.utility.ListingConstants;
 import com.google.sps.utility.ValidateInput;
 
+
```
Comment (J-coder13) [inline src/main/java/com/google/sps/servlets/CreateListing.java:31]: Extra space?

File: src/main/java/com/google/sps/servlets/CreateUser.java [hunk 3]
```diff
@@ -87,10 +87,10 @@ private Entity createUserEntity(HttpServletRequest request, String userEmail) {
     newUserEntity.setProperty("email", userEmail);
     newUserEntity.setProperty("username", "");
     newUserEntity.setProperty("bio", "");
-    newUserEntity.setProperty("createdListingKeys", "");
-    newUserEntity.setProperty("upvotedListingKeys", "");
-    newUserEntity.setProperty("downvotedListingKeys", "");
-    
+    newUserEntity.setProperty("createdListingKeys", " ");
+    newUserEntity.setProperty("upvotedListingKeys", " ");
+    newUserEntity.setProperty("downvotedListingKeys", " ");
```
Comment (J-coder13) [inline src/main/java/com/google/sps/servlets/CreateUser.java:92]: What's the reason for this change?

File: src/main/java/com/google/sps/utility/AuthenticationUtility.java (hunk 1 above)
Comment (zrideaux) [reply src/main/java/com/google/sps/utility/AuthenticationUtility.java:116]: Oops, it's actually a misspelling of deprecated, meaning that it's still there if it needs to be used, but it's best to avoid it when possible. Basically I just didn't want to delete it from the code *just* yet :)

> (chiefly of a software feature) be usable but regarded as obsolete and best avoided, typically due to having been superseded.

File: src/main/java/com/google/sps/servlets/CreateListing.java (hunk 2 above)
Comment (zrideaux) [reply src/main/java/com/google/sps/servlets/CreateListing.java:31]: Yep, removed!

File: src/main/webapp/scripts/authentication.js [hunk 4]
```diff
@@ -1,100 +1,50 @@
...
-    GoogleAuth = gapi.auth2.init({client_id: CLIENT_ID})
-    .then(() => {
-      renderButton();
-    });
-  });
-}
+// Run authenticate function on page load. 
+window.onload = function() {
+  authenticate();
+};
```
Comment (kev1n80) [inline src/main/webapp/scripts/authentication.js:5]: I think it would be best if we made this script modular in order to use the authenticate function for different scenarios, such as checking if the user is logged in before calling any user specific script in the user.html page.
If this script is modular, this function may not run.

File: src/main/java/com/google/sps/servlets/CreateUser.java (hunk 3 above)
Comment (zrideaux) [reply src/main/java/com/google/sps/servlets/CreateUser.java:92]: In datastore blank strings are retrieved as "null". While it would certainly be possible to do a check to each time we retrieve one of these properties, I figured it would save time and remove the potential for mistakes later if I just made the initial value a space

File: src/main/java/com/google/sps/servlets/Authentication.java [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+    final String LOGIN_LINK_KEY = "loginLink";
+    Gson gson = new Gson();
+
+    // Check whether or not current user is logged in.
+    if (userService.isUserLoggedIn()) {
+      // Create information to return to front end.
+      String userIsLoggedIn = "true";
+      String userIsAdmin = String.valueOf(userService.isUserAdmin());
+      String userEmail = userService.getCurrentUser().getEmail();
+      String userEntityStatus = "";
```
[the remark this answers is not in the document]
Comment (zrideaux) [reply src/main/java/com/google/sps/servlets/Authentication.java:64]: Good point! Updated

File: src/main/java/com/google/sps/servlets/Authentication.java [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+
+      String urlToRedirectToAfterUserLogsOut = "/";
+      String logoutLink = userService.createLogoutURL(
+          urlToRedirectToAfterUserLogsOut);
```
[the remark this answers is not in the document]
Comment (zrideaux) [reply src/main/java/com/google/sps/servlets/Authentication.java:68]: Sure! My main mentality was that it might be easier to change later with this variable if we need to, but it doesn't add anything to the code in its current state so I'll remove them!

File: src/main/java/com/google/sps/utility/AuthenticationUtility.java (hunk 1 above)
Comment (J-coder13) [reply src/main/java/com/google/sps/utility/AuthenticationUtility.java:116]: got it

File: src/main/java/com/google/sps/servlets/CreateListing.java (hunk 2 above)
Comment (J-coder13) [reply src/main/java/com/google/sps/servlets/CreateListing.java:31]: Nice!

File: src/main/java/com/google/sps/servlets/Authentication.java [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+    final String USER_EMAIL_KEY = "userEmail";
+    final String USER_ENTITY_STATUS_KEY = "userEntityStatus";
+    final String LOGOUT_LINK_KEY = "logoutLink";
+    final String LOGIN_LINK_KEY = "loginLink";
+    Gson gson = new Gson();
+
+    // Check whether or not current user is logged in.
+    if (userService.isUserLoggedIn()) {
+      // Create information to return to front end.
+      String userIsLoggedIn = "true";
```
[the remark this answers is not in the document]
Comment (zrideaux) [reply src/main/java/com/google/sps/servlets/Authentication.java:61]: Storing it as a string so it can be placed in the HashMap above (`HashMap<String, String>`)

File: src/main/java/com/google/sps/utility/AuthenticationUtility.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (zrideaux) [reply src/main/java/com/google/sps/utility/AuthenticationUtility.java:116]: Yep, just wanted to keep it around until the HTML files were updated!

File: src/main/java/com/google/sps/utility/AuthenticationUtility.java [hunk 8]
```diff
@@ -139,13 +140,19 @@ public Entity getCurrentUserEntity(DatastoreService datastore,
...
+  }
+
   /**
    * Retrieves the email associated with the currently logged in user.
    *
    * @param request an http request to the servlet, should include "idtoken"
    *     parameter
    */
-  public String getCurrentUserEmail(HttpServletRequest request) throws Exception {
+  public static String getCurrentUserEmail(HttpServletRequest request) throws Exception {
```
[the remark this answers is not in the document]
Comment (zrideaux) [reply src/main/java/com/google/sps/utility/AuthenticationUtility.java:155]: Thank you!

File: src/main/java/com/google/sps/servlets/CreateUser.java (hunk 3 above)
Comment (J-coder13) [reply src/main/java/com/google/sps/servlets/CreateUser.java:92]: ahh ok. do you think that that will cause any problems when we try to make a list out of those keys?

File: src/main/webapp/scripts/authentication-legacy.js [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+  console.log(googleUser.getBasicProfile().getEmail());
+  
+  // Do a quick front end check to make sure they're signed in.
+  // If they aren't, you might want to do something else like hide
+  //     HTML elements or that kind of thing.
+  if (GoogleAuth.isSignedIn.get()) {
+    // Make sure you also do a check to see they're the right user.
+    fetch('/servlet-name?idtoken=' + idToken);
+  }
+}
```
[the remark this answers is not in the document]
Comment (zrideaux) [reply src/main/webapp/scripts/authentication-legacy.js:100]: Will do after HTML files are updated!

File: src/main/java/com/google/sps/servlets/CreateUser.java (hunk 3 above)
Comment (zrideaux) [reply src/main/java/com/google/sps/servlets/CreateUser.java:92]: From my testing I don't think it will be an issue. If it does cause problems, we can use [`String.trim()`](https://www.w3schools.com/java/ref_string_trim.asp) before using `String.split(" ")` when making the list.

File: src/main/java/com/google/sps/data/User.java [hunk 10]
```diff
@@ -31,4 +36,23 @@ public User(String bio, String email, String username,
...
+   */
+  public static Entity createUserEntity(String userEmail) {
+    // Create User entity and add to datastore
+    Entity newUserEntity = new Entity("User");
+    newUserEntity.setProperty("email", userEmail);
+    newUserEntity.setProperty("username", "");
+    newUserEntity.setProperty("bio", "");
+    newUserEntity.setProperty("createdListingKeys", " ");
+    newUserEntity.setProperty("upvotedListingKeys", " ");
+    newUserEntity.setProperty("downvotedListingKeys", " ");
```
[the remark this answers is not in the document]
Comment (zrideaux) [reply src/main/java/com/google/sps/data/User.java:54]: See [this thread](https://github.com/zrideaux/capstone/pull/37#discussion_r465776990)

File: src/main/java/com/google/sps/servlets/Authentication.java [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+      authenticationInfo.put(USER_IS_LOGGED_IN_KEY, userIsLoggedIn);
+      authenticationInfo.put(USER_IS_ADMIN_KEY, userIsAdmin);
+      authenticationInfo.put(USER_EMAIL_KEY, userEmail);
+      authenticationInfo.put(USER_ENTITY_STATUS_KEY, userEntityStatus);
+      authenticationInfo.put(LOGOUT_LINK_KEY, logoutLink);
+    } else {
+      // Create information to return to front end.
+      String userIsLoggedIn = "false";
+      String userIsAdmin = "false";
+      String urlToRedirectToAfterUserLogsIn = "/";
```
[the remark this answers is not in the document]
Comment (zrideaux) [reply src/main/java/com/google/sps/servlets/Authentication.java]: Since placing some of the auth info variables directly in the `put` function would (in my opinion) decrease readability, I'd like to opt to keep all of their declarations consistent. I believe the extra space is insignificant enough to favor readability and consistency
